### PR TITLE
Decouple `service_provider_controller` from db implementation

### LINKED ISSUE #504: Decouple database implmentation in application layer
### Overview
Use `service_provider_controller` as an example to decouple database implementation.

### Action Items
- Pull out database implementation from `service_provider_controller` into class as an example.
- Show benefits of refactor using different database implementations in tests. 

### Resources/Instructions
- Move database implementation into it's own class
- Set `service_provider_controller`to reference this class
- Add tests.

### Things to Consider
At the time of writing this issue, all models are structured in an Object Relational Mapping (ORM) pattern. I suggest we use a Domain Driven Design (DDD) pattern to account for variability in our database schemas. Separating application and data layer logic could be the first step in that direction.

### PR DESCRIPTION
**Pull request by ju1es**
## Problem
Difficult to test controllers. 

Closes #504 

## Solution
- Use `service_provider_controller` as an example.
- Pull out db implementation into "repository" class.
- Add tests.

## Things To Consider
- Live instance of db still coupled to `service_provider_controller`. Hoping to use `Flask-Injector` in a separate PR.
- Need to add test runner to build.

### CODE REVIEW

File: api/openapi_server/controllers/service_provider_controller.py [hunk 1]
```diff
@@ -1,15 +1,12 @@
...
-from openapi_server.models.api_response import ApiResponse  # noqa: E501
+# Local
 from openapi_server.models.service_provider import ServiceProvider  # noqa: E501
-from openapi_server.models.service_provider_with_id import ServiceProviderWithId
-from openapi_server.models import database as db
-from sqlalchemy.orm import Session
+from openapi_server.repositories.service_provider_repository import HousingProviderRepository
 
-db_engine = db.DataAccessLayer.get_engine()
+housing_provider_repository = HousingProviderRepository()
```
Comment (ju1es) [inline api/openapi_server/controllers/service_provider_controller.py:9]: would like to do dependency injection with this using the lib `dependency_injector` but need to setup containers. figured i'd save that for a separate PR.

File: api/openapi_server/repositories/service_provider_repository.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                    db.HousingProgramServiceProvider.id == provider_id)
+        if query.first() != None:
+            query.update(provider)
+            session.commit()
+            provider["id"] = provider_id
+            result = ServiceProviderWithId.from_dict(provider)
+        session.close()
+        return result
+    
+    def _get_session(self):
```
Comment (ju1es) [inline api/openapi_server/repositories/service_provider_repository.py:125]: for overriding in testing

File: api/requirements.txt [hunk 3]
```diff
@@ -33,5 +33,5 @@ SQLAlchemy==1.4.36
 swagger-ui-bundle==0.0.9
 typing_extensions==4.5.0
 urllib3==1.26.9
-Werkzeug==2.2.3
+Werkzeug==2.3.3
```
Comment (ju1es) [inline api/requirements.txt:36]: was getting errors about dependencies when installing

File: api/test-requirements.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-pytest~=4.6.7 # needed for python 2.7+3.4
-pytest-cov>=2.8.1
-pytest-randomly==1.2.3 # needed for python 2.7+3.4
-Flask-Testing==0.8.0
+pytest==7.3.1
+pytest-cov==4.0.0
+pytest-randomly==3.12.0 
+Flask-Testing==0.8.1
```
Comment (ju1es) [inline api/test-requirements.txt:4]: was getting errors about dependencies when installing as well. why does the pytest version need to be considered for python 2.7?

**Comment by agosmou** [comment]
@ju1es 

Ill take a look at this tomorrow 6/1/23

**Comment by erikguntner** [comment]
Looks good @ju1es!

**Review by erikguntner — APPROVED** [review]

**Comment by ju1es** [comment]
rebased and squashed

**Pull request merged**